komodo/lint_maturity.py

```diff
@@ -0,0 +1,206 @@
+    count_tag_maturity = {tag: len(dict_tag_maturity[tag]) for tag in dict_tag_maturity}
+    invalid_tags_count = sum(count_tag_maturity[tag] for tag in invalid_tags)
+    return invalid_tags_count
+
+
+def warning_message_packages(release_basename, dict_tag_maturity):
+    warning_msg = ""
+    if len(dict_tag_maturity["exception"]) > 0:
+        warning_msg += release_basename + ", exception list of packages:\n"
+        warning_msg += "\t" + str(dict_tag_maturity["exception"]) + " \n"
+    return warning_msg
+
+
+def warning_message_release(release_basename, release_version):
+    warning_msg = ""
+    if release_version == "exception":
+        warning_msg += release_basename + " is in the exception list of releases.\n"
+    return warning_msg
+
+
+def exit_message_packages(
+    release_basename, release_version, count_tag_invalid, dict_tag_maturity
+):
+    exit_msg = ""
+    exit_msg += (
+        release_basename
+        + " has "
+        + str(count_tag_invalid)
+        + " packages with invalid maturity tag.\n"
+    )
+    for tag in INVALID_TAGS[release_version]:
+        if len(dict_tag_maturity[tag]) > 0:
+            exit_msg += (
+                "\tTag " + tag + " packages: " + str(dict_tag_maturity[tag]) + "\n"
+            )
+    return exit_msg
+
+
+def print_warning_message(system_warning_msg):
+    if system_warning_msg != "":
+        warnings.warn(system_warning_msg, UserWarning)
+
+
+def print_system_exit_message(system_exit_msg):
+    if system_exit_msg != "":
+        raise SystemExit(system_exit_msg)
+
+
+def run(files_to_lint, tag_exceptions):
+
+    system_exit_msg = ""
+    system_warning_msg = ""
+
+    for file_to_lint in files_to_lint:
+
+        release_basename = os.path.basename(file_to_lint)
+        release_version = get_release_version(
+            release_basename, tag_exceptions["release"]
+        )
+        system_warning_msg += warning_message_release(release_basename, release_version)
+
+        if release_version == "invalid":
+            system_exit_msg += (
+                release_basename + " is incompatible with version name.\n"
+            )
+        else:
+            packages_dict = read_yaml_file(file_path=file_to_lint)
+            dict_tag_maturity = get_packages_info(
+                packages_dict, tag_exceptions["package"]
+            )
+            count_tag_invalid = count_invalid_tags(
+                dict_tag_maturity, INVALID_TAGS[release_version]
+            )
+
+            system_warning_msg += warning_message_packages(
+                release_basename, dict_tag_maturity
+            )
+
+            if count_tag_invalid > 0:
+                system_exit_msg += exit_message_packages(
+                    release_basename,
+                    release_version,
+                    count_tag_invalid,
+                    dict_tag_maturity,
+                )
+
+    print_warning_message(system_warning_msg)
+    print_system_exit_message(system_exit_msg)
+
+
+def main():
```

Comment:
You have spotted a new function to be defined, thanks! :-) 
Now we are smaller functions in main() and unit tests for all of them.
An integration test is in run(), so I believe we do not need to add a integration test to main().
Would you agree with that?

Reply:
Not quite, although most of your logic is contained in the `run` function, and your code is well tested, it would be good to have an integration test of the full workflow as well, so all your code is actually tested. Theoretically I could now delete everything in `main` and the tests would still be green.

Reply:
If you would also like to add a unit test to `main`, it could look something like this:
```
def test_main(monkeypatch):
    parser_mock = mock.Mock()
    args = mock.Mock()
    args.tag_exceptions = "an_exception"
    parser_mock.parse_args.return_value = args
    get_parser_mock = mock.Mock(return_value=parser_mock)
    monkeypatch.setattr(lint_maturity, "get_parser", get_parser_mock)

    tag_exceptions_mock = mock.Mock(return_value=["list", "of", "exceptions"])
    monkeypatch.setattr(lint_maturity, "define_tag_exceptions", tag_exceptions_mock)

    files_to_lint_mock = mock.Mock(return_value=["list", "of", "files"])
    monkeypatch.setattr(lint_maturity, "get_files_to_lint", files_to_lint_mock)

    run_mock = mock.Mock()
    monkeypatch.setattr(lint_maturity, "run", run_mock)

    assert tag_exceptions_mock.called_once_with("an_exception")
    assert files_to_lint_mock.called_once_with(args)
    assert run_mock.called_once_with(["list", "of", "files"], ["list", "of", "exceptions"])

    lint_maturity.main()
```

Reply:
This _mock_ concept is quite new for me. Many thanks. 
I tried to implement this test. Please let me know if you have concerns on that.